text/1010-global-vars.rst

```diff
@@ -0,0 +1,179 @@
+power.  Consider the ubiquitous concept of authentication, where there is
+a notion of "current user" that is used to filter query results or make other
+decisions.  Currently, such data must be passed as an explicit argument to
+queries and there is no way to pass such context to schema expressions at all.
+
+
+Specification
+=============
+
+A global variable must be declared before it can be used in any expression.
+
+CREATE GLOBAL
+-------------
+
+Define a new global variable.
+
+Synopsis::
+
+    [ WITH <with-item> [, ...] ]
+    CREATE [ REQUIRED ] [ MULTI | SINGLE] GLOBAL <name> -> <type> [ "{"
+        <subcommand> ; [...]
+    "}" ] ;
+
+    # Computed global variable form:
+
+    [ WITH <with-item> [, ...] ]
+    CREATE [ REQUIRED ] GLOBAL <name> := <expr> ;
+
+    # where <subcommand> is one of
+
+      SET default := <value>
+      RESET default
+      RENAME TO <newname>
+      USING <expr>
+      RESET EXPRESSION
+      CREATE ANNOTATION <annotation-name> := <value>
+      ALTER ANNOTATION <annotation-name> := <value>
+      DROP ANNOTATION <annotation-name>
+
+The ``<type>`` of non-computed globals must be a primitive type, i.e. it
+*must not* be an object type or a collection of object types.  This is because
+object type variables will require some form of referential integrity checks
+like links do (i.e. the ``on target delete`` spec), which complicates the
+implementation considerably.  Computed globals can have any type, and so it
+is possible to define an object type global in terms of its ``id``, e.g::
+
+    CREATE GLOBAL user_id -> uuid;
+    CREATE GLOBAL user := (SELECT User FILTER .id = (GLOBAL user_id));
+
+If a global is declared as ``REQUIRED``, it *must* provide a default value, or,
+if the global is computed, the expression must be non-optional.
+
+Explicit cardinality is specified via the ``SINGLE`` or ``MULTI`` keyword,
+otherwise the cardinality is assumed to be ``SINGLE`` in non-computed globals,
+and is inferred from the expression in computed globals.
+
+
+ALTER GLOBAL
+------------
+
+Alter the definition of a global variable.
+
+Synopsis::
+
+    [ WITH <with-item> [, ...] ]
+    ALTER GLOBAL <name>
+    "{" <subcommand>; [...] "}" ;
+
+    # where <subcommand> is one of
+
+      SET default := <value>
+      RESET default
+      RENAME TO <newname>
+      USING <expr>
+      RESET EXPRESSION
+      SET REQUIRED
+      DROP REQUIRED
+      SET SINGLE
+      SET MULTI
+      CREATE ANNOTATION <annotation-name> := <value>
+      ALTER ANNOTATION <annotation-name> := <value>
+      DROP ANNOTATION <annotation-name>
+
+
+DROP GLOBAL
+-----------
+
+Remove a global variable.
+
+Synopsis::
+
+    [ WITH <with-item> [, ...] ]
+    DROP GLOBAL <name> ;
+
+
+SET GLOBAL
+----------
+
+Set the value of a non-computed global variable *in the current session*
+by evaluating the given expression.
+
+Synopsis::
+
+    SET GLOBAL <name> := <expr> ;
```

Comment:
What about setting multiple variables? Making each of them separate query (so separate roundtrip or 3 if there is no cache) is very unfortunate.

Reply:
This assumes we have `execute` with multiple statements, but we don't have as far as I remember, right?

Although, if this is mostly for REPL and this is not the main way to set session variables in client bindings, then fine.